# <span style="color:#54B1FF">Probability:</span> &nbsp; <span style="color:#1B3EA9"><b>Analytical PDF Calculations</b></span>

<br>

The easisest way to analytically calculate probabilities in Python is to use the `sf` function of `scipy.stats` distributions.

`sf` means **survival function**. Survival functions are defined as $f(u) = P(x > u)$, where $x$ is a random variable, and where $u$ is a threshold. They are called "survival" functions because they represent the probability that a random value $x$ will survive a thresholding process.

`sf` functions can be used in Python as follows:

<br>

In [1]:
from scipy import stats

In [2]:

p0 = stats.uniform.sf(1.0) # P(x > 1.0) for uniform distribution
p1 = stats.uniform.sf(0.5) # P(x > 0.5) for uniform distribution
p2 = stats.norm.sf(0)      # P(x > 0.0) for normal distribution

print('Uniform:  P(x > 1.0) = ', p0)
print('Uniform:  P(x > 0.5) = ', p1)
print('Normal:   P(x > 0.0) = ', p2)

Uniform:  P(x > 1.0) =  0.0
Uniform:  P(x > 0.5) =  0.5
Normal:   P(x > 0.0) =  0.5


<br>
<br>
<br>

Note that these probabilities correspond to the first three probability values mentioned in the previous notebook (copied here):

**Uniform**:

* The probablity that $x$ is greater than one is zero: $\hspace{5mm} P(x>1) = 0$
* The probablity that $x$ is greater than 0.5 is 0.5: $\hspace{5mm} P(x>0.5) = 0.5$

**Normal**:

* The probablity that $x$ is greater than zero is 0.5: $\hspace{5mm} P(x>0) = 0.5$
* The probablity that $x$ is less than zero is 0.5: $\hspace{5mm} P(x<0) = 0.5$

<br>
<br>

How about the final probability:   $P(x<0) = 0.5$ (normal distribution);  how can this be verified in Python?  

Probabilities of the form  $P(x<u)$ can be calculated as:

$$P(x<u) = 1 - P(x>u)$$

<br>

In Python:

<br>

In [3]:
p3 = 1 - stats.norm.sf(0)    # P(x < 0.0) for normal distribution

print('Normal:   P(x < 0.0) = ', p3)

Normal:   P(x < 0.0) =  0.5


<br>
<br>

Now that we can compute analytical probabilities in Python, we can compute less obvious probabilities for the distributions in Fig.7.1 (see previous notebook).

For example: what is the probability that $x$ is greater than 2.0 if $x$ is a random normal variable?

From Fig.7.1 it appears that approximately 5% of the distribution is greater than 2.0, but how can we know the exact value?

We can use the `sf` function as follows:

<br>

In [4]:
p4 = stats.norm.sf(2.0)    # P(x > 2.0) for normal distribution

print('Normal:   P(x > 2.0) = ', p4)

Normal:   P(x > 2.0) =  0.022750131948179195


<br>
<br>

This result suggests that only about 2.3% of the distribution lies above 2.0.  In other words, there is only a 2.3% probability that $x$ is greater than 2.

How about the $\chi^2$ distribution? From Fig.7.1 it looks like approximately 40% of the distribution lies above 5. We can calculate the actual amount as follows:

<br>

In [5]:
p5 = stats.chi2.sf(5.0, 4)    # P(x > 5.0) for chi-squared distribution with degrees of freedom = 4

print('Chi-squared (v=4):   P(x > 5.0) = ', p5)

Chi-squared (v=4):   P(x > 5.0) =  0.2872974951836458


<br>

This result shows that only about 28.7% of the distribution lies above 5.

All of these `sf` results are **analytical**.  In other words, the `sf` functions represent real mathematical functions.

How can we know that the `sf` results are accurate?

One way is to use numerical simulation. This is discussed in the next notebook.